File: python/tsp.py
import sys
import time


def get_shortest_path(cur, visited):
    if visited == (1 << n) - 1:
        return w[cur][0]

    if dp[cur][visited] >= 0:
        return dp[cur][visited]

    ret = 9876543210

    for i in range(n):
        if (visited & (1 << i)) != 0:
            continue

        if w[cur][i] == 0:
            continue

        ret = min(ret, w[cur][i] + get_shortest_path(i, visited + (1 << i)))

    dp[cur][visited] = ret
    return ret


for k in range(2, 24):
    f = open("../data/" + str(k) + ".txt","r")
    
    n = int(f.readline())
    w = [[]] * n
    dp = [[-1] * (1 << n) for _ in range(n)]
    for i in range(n):
        w[i] = list(map(int, f.readline().replace("\n", "").split()))
        
    start = time.time()
    print("N = " + str(n) + " :: " + str(get_shortest_path(0, 1)))
    finish = time.time()
    print("Execution Time : " + str(finish-start) + "\n")

File: cpp/tsp.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <cstdio>
#include <algorithm>
#include <functional>
#include <iostream>
#include <map>
#include <vector>
#include <cstring>
#include <set>
#include <cmath>
#include <time.h>
#include <limits.h>

using namespace std;

//typedef int long long;
typedef long long LL;
typedef vector<LL> VLL;

#define MAX_NODE 23

using namespace std;

int n;
int cost[MAX_NODE][MAX_NODE];
int dp[MAX_NODE][1 << MAX_NODE];

int tsp2(int here, int visited) {
	if (visited == (1 << n) - 1) {
		return cost[here][0];
	}

	int& ret = dp[here][visited];
	if (ret >= 0) {
		return ret;
	}
	ret = INT_MAX;

	for (int next = 0; next < n; next++) {
		if (visited & (1 << next)) {
			continue;

		}

		int candidate_dist = cost[here][next] + tsp2(next, visited + (1 << next));
		ret = min(ret, candidate_dist);
	}

	return ret;
}

int main() {
	FILE* result = NULL;
	result = fopen("result.txt", "w");

	for (int i = 0; i<23; i++) {
		memset(cost, 0, sizeof(cost));
		memset(dp, -1, sizeof(dp));
		if (n == MAX_NODE + 1) break;
		FILE* fp = NULL;
		char file_name[100];
		sprintf(file_name, "%d", i+2);

		strcat(file_name, ".txt");

		fp = fopen(file_name, "r");

		fscanf(fp, "%d", &n);

		clock_t st = clock();
		for (int i = 0; i < n; i++)
			for (int j = 0; j < n; j++)
				fscanf(fp, "%d", &cost[i][j]);


		int ans = tsp2(0, 1);
		clock_t fin = clock();

		printf("Minimum Cost When N=%d :: %d. Execution time : %d ms\n", n, ans, fin - st);
		fprintf(result, "Minimum Cost When N=%d :: %d. Execution time : %d ms\n", n, ans, fin - st);

		fclose(fp);
	}
	fclose(result);
	return 0;
}